kivy/tests/test_lang_pre_process_and_post_process.py

```diff
@@ -0,0 +1,108 @@
+from kivy.factory import Factory
+from kivy.properties import ObjectProperty, NumericProperty
+
+
+KV_CODE = '''
+<EventCounter>:
+    on_kv_pre: self.on_kv_pre_from_kv()  # This line won't be excuted
+    on_kv_post: self.on_kv_post_from_kv()
+<TestBoxLayout>:
+    Label:
+        id: label
+        text: textinput.text
+    TextInput:
+        id: textinput
+    Button:
+        id: button
+        on_press: self.text = 'pressed'
+'''
+KV_FILENAME = 'test_kv_filename'
+
+
+def setUpModule():
+    Builder.load_string(KV_CODE, filename=KV_FILENAME)
+
+
+def tearDownModule():
+    Builder.unload_file(KV_FILENAME)
+    u = Factory.unregister
+    u('EventCounter')
+    u('TestBoxLayout')
+
+
+class LangTestCase(unittest.TestCase):
+
+    def test_event_count(self):
+        testcase = self
+        ae = self.assertEqual
+        NP = NumericProperty
+
+        class EventCounter(Factory.Widget):
+            _n_post_kv = NP(0)
+            _n_pre = NP(0)
+            _n_post = NP(0)
+
+            def on_kv_pre(self):
+                self._n_pre += 1
+                ae(self._n_pre, 1)
+                ae(self._n_post_kv, 0)
+                ae(self._n_post, 0)
+
+            def on_kv_post(self):
+                self._n_post += 1
+                self.assert_all_events_are_dispatched()
+
+            def on_kv_pre_from_kv(self):
+                testcase.fail('This method is supposed to not being called')
+
+            def on_kv_post_from_kv(self):
+                self._n_post_kv += 1
+                ae(self._n_pre, 1)
+                ae(self._n_post_kv, 1)
+                ae(self._n_post, 0)
+
+            def assert_all_events_are_dispatched(self):
+                ae(self._n_pre, 1)
+                ae(self._n_post_kv, 1)
+                ae(self._n_post, 1)
+
+        EventCounter().assert_all_events_are_dispatched()
+
+        root = Builder.load_string(textwrap.dedent('''
+        EventCounter:
+            EventCounter:
+                id: ec
+        '''))
+        root.assert_all_events_are_dispatched()
+        root.ids.ec.assert_all_events_are_dispatched()
+
+    def test_rule_is_applied_on_kv_post(self):
+        tc = self
+
+        class TestBoxLayout(Factory.BoxLayout):
+            def on_kv_post(self):
+                ids = self.ids
+                tc.assertIn('textinput', ids)
+                tc.assertIn('label', ids)
+                tc.assertIn('button', ids)
+
+                textinput = ids.textinput
+                label = ids.label
+                label.text = 'A'
+                textinput.text = 'B'
+                tc.assertEqual(label.text, 'B')
+
+                button = ids.button
+                button.text = ''
+                button.dispatch('on_press')
+                tc.assertEqual(button.text, 'pressed')
+
+                tc.assertTrue(self.parent is not None)
+
+        root = Builder.load_string(textwrap.dedent('''
+        Widget:
+            TestBoxLayout:
+        '''))
```

Comment:
There should be tests for more complex situations. E.g. the code above:

```
<MyWidget@Widget>:
    name: 'matt'
    on_kv_post: self.height = 40
    Label:
        text: print('other') or 'hello'

<MyOtherWidget@MyWidget>:
    name: 'other matt'
    on_kv_post: self.height = 41
    height: print('something') or 90
    Label:
        text: print('other2') or 'hello2'

BoxLayout:
    MyOtherWidget:
        on_kv_post: self.height = 42
```

should set `height` three times, and all should happen *after* all the print statements.

Reply:
Hi, I think your code above will set `height` **four** times, which are:

- `on_kv_post: self.height = 40`
- `on_kv_post: self.height = 41`
- `height: print('something') or 90`
- `on_kv_post: self.height = 42`

But the more important question is: Are you sure that "all should happen **after** all the print statements"? I think there is a possibility of `height: print('something') or 90` will be excuted **before** other print statements.

Reply:
I meant the latter - the three  `on_kv_post` height assignments should happen after all the other print statements.

Reply:
ah ok got it